src/ngx_http_lua_shdict.c

```diff
@@ -1482,107 +1582,794 @@ ngx_http_lua_ffi_shdict_store(ngx_shm_zone_t *zone, int op, u_char *key,
 
-        goto replace;
-    }
+    dd("shdict lookup returned %d", (int) rc);
 
-    if (op & NGX_HTTP_LUA_SHDICT_ADD) {
+    if (rc == NGX_DONE) {
+        /* exists but expired */
 
-        if (rc == NGX_OK) {
-            ngx_shmtx_unlock(&ctx->shpool->mutex);
-            *errmsg = "exists";
-            return NGX_DECLINED;
-        }
+        if (sd->value_type != SHDICT_TLIST) {
+            /* TODO: reuse when length matched */
 
-        if (rc == NGX_DONE) {
-            /* exists but expired */
+            ngx_log_debug0(NGX_LOG_DEBUG_HTTP, ctx->log, 0,
+                           "lua shared dict push: found old entry and value "
+                           "type not matched, remove it first");
 
-            dd("go to replace");
-            goto replace;
-        }
+            ngx_queue_remove(&sd->queue);
 
-        /* rc == NGX_DECLINED */
+            node = (ngx_rbtree_node_t *)
+                        ((u_char *) sd - offsetof(ngx_rbtree_node_t, color));
 
-        dd("go to insert");
-        goto insert;
-    }
+            ngx_rbtree_delete(&ctx->sh->rbtree, node);
 
-    if (rc == NGX_OK || rc == NGX_DONE) {
+            ngx_slab_free_locked(ctx->shpool, node);
 
-        if (value_type == LUA_TNIL) {
-            goto remove;
+            dd("go to init_list");
+            goto init_list;
         }
 
-replace:
-
-        if (str_value_buf && str_value_len == (size_t) sd->value_len) {
-
-            ngx_log_debug0(NGX_LOG_DEBUG_HTTP, ctx->log, 0,
-                           "lua shared dict set: found old entry and value "
-                           "size matched, reusing it");
-
-            ngx_queue_remove(&sd->queue);
-            ngx_queue_insert_head(&ctx->sh->queue, &sd->queue);
+        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, ctx->log, 0,
+                       "lua shared dict push: found old entry and value "
+                       "type matched, reusing it");
 
-            sd->key_len = (u_short) key_len;
+        sd->expires = 0;
 
-            if (exptime > 0) {
-                tp = ngx_timeofday();
-                sd->expires = (uint64_t) tp->sec * 1000 + tp->msec
-                              + (uint64_t) exptime;
+        /* free list nodes */
 
-            } else {
-                sd->expires = 0;
-            }
+        queue = ngx_http_lua_shdict_get_list_head(sd, key.len);
 
-            sd->user_flags = user_flags;
+        for (q = ngx_queue_head(queue);
+             q != ngx_queue_sentinel(queue);
+             q = ngx_queue_next(q))
+        {
+            /* TODO: reuse matched size list node */
+            lnode = ngx_queue_data(q, ngx_http_lua_shdict_list_node_t, queue);
+            ngx_slab_free_locked(ctx->shpool, lnode);
+        }
 
-            sd->value_len = (uint32_t) str_value_len;
+        ngx_queue_init(queue);
 
-            dd("setting value type to %d", value_type);
+        ngx_queue_remove(&sd->queue);
+        ngx_queue_insert_head(&ctx->sh->queue, &sd->queue);
```

Comment:
Because we have two different queues here, list-typed values and the global LRU queue, maybe it's the time to rename `ctx->sh->queue` to `ctx->sh->lru_queue` to be clearer? Frankly I stumbled on these lines and got confused with these two different queues mixed together in the code base ;)